File: numerics/src/diagnostics.py
"""Diagnostics: fit summaries, model comparison, true-vs-fitted crosstabs."""
from __future__ import annotations

import numpy as np
import pandas as pd

from .hmm_em import HMMFit
from .mixture_hmm_em import MixtureHMMFit
from .simulate import REFINED_LABELS


def fit_summary_rows(fit: HMMFit, T: int) -> list[dict]:
    """One row per fitted state: mean, sd, posterior occupation, logLik, BIC, ICL."""
    occ = fit.gamma.mean(axis=0)
    return [
        {
            "K": fit.K,
            "state": j + 1,
            "mean": fit.means[j],
            "sd": fit.sds[j],
            "posterior_occupation": occ[j],
            "logLik": fit.loglik,
            "BIC": fit.bic(T),
            "ICL": fit.icl(T),
        }
        for j in range(fit.K)
    ]


def crosstab_true_vs_fitted(refined: np.ndarray, fit: HMMFit,
                            use: str = "viterbi",
                            proportions: bool = False) -> pd.DataFrame:
    """Crosstab of true refined states {1a, 1b, 2} vs fitted state labels.

    use : "viterbi" (decoded path) or "posterior" (argmax of gamma).
    """
    fitted = fit.viterbi if use == "viterbi" else fit.gamma.argmax(axis=1)
    tab = pd.crosstab(
        pd.Series(REFINED_LABELS[refined], name="true_refined"),
        pd.Series(fitted + 1, name=f"fitted_state_K{fit.K}"),
    )
    if proportions:
        tab = tab.div(tab.sum(axis=1), axis=0)
    return tab


def model_comparison(fits: dict[int, HMMFit], T: int) -> dict:
    """Delta logLik and Delta BIC between K=3 and K=2 (plus all raw values)."""
    out = {}
    for K, fit in fits.items():
        out[f"logLik_K{K}"] = fit.loglik
        out[f"BIC_K{K}"] = fit.bic(T)
        out[f"ICL_K{K}"] = fit.icl(T)
    if 2 in fits and 3 in fits:
        out["delta_logLik_3_minus_2"] = fits[3].loglik - fits[2].loglik
        out["delta_BIC_3_minus_2"] = fits[3].bic(T) - fits[2].bic(T)
    return out


def mixture_fit_summary_rows(fit: MixtureHMMFit, T: int) -> list[dict]:
    """One row per state-component of a mixture-emission HMM."""
    occ = fit.gamma.mean(axis=0)
    rows = []
    for k in range(fit.K):
        for m in range(fit.M):
            rows.append({
                "K": fit.K,
                "M": fit.M,
                "state": k + 1,
                "component": m + 1,
                "component_weight": fit.weights[k, m],
                "mean": fit.means[k, m],
                "sd": fit.sds[k, m],
                "state_occupation": occ[k],
                "logLik": fit.loglik,
                "BIC": fit.bic(T),
            })
    return rows


def one_step_residuals_gaussian(y: np.ndarray, fit: HMMFit) -> np.ndarray:
    """Approximate one-step standardized prediction residuals."""
    pred = fit.pi.copy()
    residuals = np.zeros(len(y))
    for t, yt in enumerate(y):
        mean = float(np.sum(pred * fit.means))
        second = float(np.sum(pred * (fit.sds ** 2 + fit.means ** 2)))
        var = max(second - mean ** 2, 1e-10)
        residuals[t] = (yt - mean) / np.sqrt(var)
        dens = np.exp(-0.5 * ((yt - fit.means) / fit.sds) ** 2)
        dens /= np.sqrt(2.0 * np.pi) * fit.sds
        filt = pred * np.maximum(dens, 1e-300)
        filt /= filt.sum()
        pred = filt @ fit.A
    return residuals


def residual_acf(residuals: np.ndarray, max_lag: int = 20) -> pd.DataFrame:
    """Sample autocorrelation function for lags 1..max_lag."""
    x = residuals - residuals.mean()
    denom = float(np.dot(x, x))
    rows = []
    for lag in range(1, max_lag + 1):
        acf = float(np.dot(x[:-lag], x[lag:]) / denom) if denom > 0 else np.nan
        rows.append({"lag": lag, "acf": acf})
    return pd.DataFrame(rows)

File: numerics/R/diagnostics.R
# Numerical summaries and observable diagnostics for fitted HMMs.

gaussian_fit_summary <- function(fit, T) {
  data.frame(
    K = fit$K,
    state = seq_len(fit$K),
    mean = fit$means,
    sd = fit$sds,
    posterior_occupation = colMeans(fit$posterior),
    logLik = fit$log_likelihood,
    BIC = fit_bic(fit, T),
    ICL = fit_icl(fit, T),
    check.names = FALSE
  )
}

mixture_fit_summary <- function(fit, T) {
  rows <- vector("list", fit$K * fit$M)
  index <- 1L
  occupations <- colMeans(fit$posterior)
  for (state in seq_len(fit$K)) {
    for (component in seq_len(fit$M)) {
      rows[[index]] <- data.frame(
        K = fit$K,
        M = fit$M,
        state = state,
        component = component,
        component_weight = fit$weights[state, component],
        mean = fit$means[state, component],
        sd = fit$sds[state, component],
        state_occupation = occupations[[state]],
        logLik = fit$log_likelihood,
        BIC = fit_bic(fit, T)
      )
      index <- index + 1L
    }
  }
  do.call(rbind, rows)
}

crosstab_true_fitted <- function(refined_state, fit, use = c("viterbi", "posterior")) {
  use <- match.arg(use)
  fitted <- if (use == "viterbi") fit$viterbi else max.col(fit$posterior)
  table(
    true_refined = factor(
      TRUE_REFINED_LABELS[refined_state],
      levels = TRUE_REFINED_LABELS
    ),
    fitted_state = factor(fitted, levels = seq_len(fit$K))
  )
}

one_step_residuals_gaussian <- function(y, fit) {
  prediction <- fit$initial
  residuals <- numeric(length(y))
  for (time in seq_along(y)) {
    prediction_mean <- sum(prediction * fit$means)
    prediction_second <- sum(prediction * (fit$sds^2 + fit$means^2))
    prediction_variance <- max(prediction_second - prediction_mean^2, 1e-10)
    residuals[[time]] <- (y[[time]] - prediction_mean) / sqrt(prediction_variance)
    density <- stats::dnorm(y[[time]], mean = fit$means, sd = fit$sds)
    filtered <- normalize_probability(prediction * pmax(density, 1e-300))
    prediction <- as.numeric(filtered %*% fit$transition)
  }
  residuals
}

residual_acf <- function(residuals, max_lag = 20L) {
  centered <- residuals - mean(residuals)
  denominator <- sum(centered^2)
  data.frame(
    lag = seq_len(max_lag),
    acf = vapply(seq_len(max_lag), function(lag) {
      if (denominator <= 0) {
        return(NA_real_)
      }
      sum(
        centered[seq_len(length(centered) - lag)] *
          centered[seq.int(lag + 1L, length(centered))]
      ) / denominator
    }, numeric(1))
  )
}
